src/proof.rs

```diff
@@ -0,0 +1,170 @@
+
+use crypto::digest::Digest;
+
+use hashable::{ Hashable };
+use merkledigest::{ MerkleDigest };
+use merkletree::{ MerkleTree };
+use tree::{ Tree };
+
+/// An inclusion proof represent the fact that `value` is a member of a `MerkleTree`
+/// with root hash `root_hash`, and hash function `digest`.  
+/// A proof is a linked-list of `ProofBlock`s.  
+/// TODO: Represent a proof as a vector of ProofBlock instead of a linked-list?
+pub struct Proof<D, T> {
+    pub digest: D,
+    pub root_hash: Vec<u8>,
+    pub block: ProofBlock,
+    pub value: T
+}
+
+impl <D, T> Proof<D, T> where D: Digest + Clone, T: Hashable + Clone {
+
+    // TODO: Refactor this mess (@romac)
+    pub fn validate_against(&self, tree: &mut MerkleTree<D, T>) -> bool {
```

Comment:
Makes a lot of sense! I went a bit too far with the verification part. I will try to address that  tomorrow.

On the other hand, does the proof generation + structure make sense to you?